### DISCUSSION TITLE
C module  in micropython

### DISCUSSION THREAD

**Discussion by micropyt** [micropython/micropython#10012] [General]
Hello, I have a problem with creating the mpy file of cmodule.
Is it possible to call another C function inside like this?

```py
STATIC mp_obj_t Timers_PWM_OFF(void)
{
  turn_PWM_OFF(); 

  return mp_obj_new_bool (true);

}

**Comment by git-n-pissed** [comment]
The only experience I have adding C functions is in `modsys.c` so they are available in the `sys` module.  Adding functions like that requires three changes to `modsys.c`

First, create the function itself.  I'm not sure how picky MicroPython is about the name of the function, but mine have always followed this naming format with the "my_function_name" being what you would change:

```
STATIC mp_obj_t mp_micropython_my_funcion_name(void) {
    return mp_obj_new_bool(true);  // Whatever you need to return
}
```

Second, define a constant function object.  I do this right below the function that was just created:
```
STATIC MP_DEFINE_CONST_FUN_OBJ_0(mp_micropython_my_function_name_obj, mp_micropython_my_function_name);
```

Third, the function needs to be mapped in `mp_module_sys_globals_table`:
```
STATIC const mp_rom_map_elem_t mp_module_sys_globals_table[] = {
    { MP_ROM_QSTR(MP_QSTR___name__), MP_ROM_QSTR(MP_QSTR_sys) },  //  I believe this line needs to be first in the map.  Put your function mappings below it.
    { MP_ROM_QSTR( MP_QSTR_my_function_name), MP_ROM_PTR(&mp_micropython_my_function_name_obj) },
    ...
```

The process for adding C functions to other modules (i.e. not `sys`) should be similar.

**Comment by jimmo** [comment]
> Is it possible to call another C function inside like this?

@micropyt You might need to provide more context. In particular, what do you mean by "creating the mpy file of cmodule" -- is this https://docs.micropython.org/en/latest/develop/natmod.html ?

Yes you can call any C function, but it will need to be statically linked into your module.

**Reply by micropyt** [reply]
> > Is it possible to call another C function inside like this?
> 
> @micropyt You might need to provide more context. In particular, what do you mean by "creating the mpy file of cmodule" -- is this https://docs.micropython.org/en/latest/develop/natmod.html ?
> 
> Yes you can call any C function, but it will need to be statically linked into your module.

Yes, I mean this case. And my question is: can I have multiple c files and one file with implementation for micropython, which I can call other C functions like the above example? Should I also include the other h files to makefile. Actually, I have a problem with this.

**Reply by jimmo** [reply]
Yes you can have multiple C files -- see the features2 example -- https://github.com/micropython/micropython/tree/master/examples/natmod/features2

Also the btree example brings in the whole Berkeley-db library.

**Reply by micropyt** [reply]
> Yes you can have multiple C files -- see the features2 example -- https://github.com/micropython/micropython/tree/master/examples/natmod/features2
> 
> Also the btree example brings in the whole Berkeley-db library.

Thank you. I have also a problem how to build for example this: https://github.com/micropython/micropython/tree/master/examples/usercmodule/cexample for stm32 Nucleo board.  There are examples of ESP everywhere ...

**Reply by jimmo** [reply]
This is a User C Module, not a dynamic native module.

See https://docs.micropython.org/en/latest/develop/cmodules.html#compiling-the-cmodule-into-micropython

There's an example there (search for "For example, building the stm32 port").

**Reply by micropyt** [reply]
> This is a User C Module, not a dynamic native module.
> 
> See https://docs.micropython.org/en/latest/develop/cmodules.html#compiling-the-cmodule-into-micropython
> 
> There's an example there (search for "For example, building the stm32 port").

Thank you.
So, I can't build it for a specific board?
Like this, it doesn't work...
micropython/ports/stm32/boards/NUCLEO_F767ZI$ make USER_C_MODULES=../../examples/usercmodule

**Reply by dlech** [reply]
How about `make -C ports/stm32 BOARD=NUCLEO_F767ZI USER_C_MODULES=examples/usercmodule`?

**Reply by micropyt** [reply]
> make -C ports/stm32 BOARD=NUCLEO_F767ZI USER_C_MODULES=examples/usercmodule

this makes only the board without the usercmodule

**Reply by dlech** [reply]
Maybe I should not have removed the `../../`? 

    make -C ports/stm32 BOARD=NUCLEO_F767ZI USER_C_MODULES=../../examples/usercmodule

**Reply by micropyt** [reply]
> Maybe I should not have removed the `../../`?
> 
> ```
> make -C ports/stm32 BOARD=NUCLEO_F767ZI USER_C_MODULES=../../examples/usercmodule
> ```
yes it works.
How can I write makefile to search for all header files?
this not working ```CFLAGS_USERMOD += -I$(USERMOD_DIR)/ ```

**Reply by micropyt** [reply]
> > Is it possible to call another C function inside like this?
> 
> @micropyt You might need to provide more context. In particular, what do you mean by "creating the mpy file of cmodule" -- is this https://docs.micropython.org/en/latest/develop/natmod.html ?
> 
> Yes you can call any C function, but it will need to be statically linked into your module.

What do you mean by statically linked?  Do you mean linked by header file?

**Reply by dlech** [reply]
> How can I write makefile to search for all header files?

You can use `CFLAGS_MOD=...` to add any addition compiler flags like `-I` for additional header file locations. (There is also `CXXFLAGS_MOD` and `LDFLAGS_MOD` for C++ flags and linker flags.)

**Reply by stinos** [reply]
Wait wasn't CFLAGS_MOD largely or even completely removed? See d6d87225585a9494093d791c807bce652e4c82d8.

**Reply by dlech** [reply]
Thanks, I must have been looking at an older version.

**Reply by jimmo** [reply]
Yes, they are now replaced with

`CFLAGS_USERMOD`, `CXXFLAGS_USERMOD`, `LDFLAGS_USERMOD`

**Reply by jimmo** [reply]
> What do you mean by statically linked? Do you mean linked by header file?

@micropyt The dynamic native module must not have any dependencies -- i.e. any code that it depends on must be statically linked into it. This is unrelated to header files.
